### PR TITLE
Add search-disabled-repos plugin.

### PR DESCRIPTION
**Pull request by megaumi**
New plugin as discussed in https://bugzilla.redhat.com/show_bug.cgi?id=1232232#c10

### CODE REVIEW

**Comment by dgoodwin** [comment]
Can one of the admins verify this patch?

**Comment by awood** [comment]
ok to test

**Comment by awood** [comment]
Would you mind updating `subscription-manager.spec` to include the new plugin file?

Should be on or around line 346

```
%{_prefix}/lib/yum-plugins/search-disabled-repos.py*
```

File: src/plugins/search-disabled-repos.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+disabled repos and searching for missing dependencies.
+To enable this functionality please set 'notify_only=0' in %s
+**********************************************************************\n""" % PLUGIN_CONF_PATH
+    conduit.info(1, msg) # yum's debuglevel 1
+
+
+def prompt_temporarily_enable_repos(conduit):
+    msg = conduit.pretty_output_restring() + """\n**********************************************************************
+Dependency resolving failed due to missing dependencies.
+Some repositories on your system are disabled, yum can enable them
```
Comment (awood) [inline src/plugins/search-disabled-repos.py]: How about "Some repositories on your system are disabled, but yum can enable them..."?  Adding the "but" avoids a run-on sentence.

File: src/plugins/search-disabled-repos.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    prompt = 'Enable all repositories and try again? [y/N]: '
+    return conduit.promptYN(msg, prompt=prompt)
+
+
+def prompt_permanently_enable_repos(conduit, repos):
+    msg = """*******************************************************************
+Dependency resolving was successful thanks to enabling these repositories:
+*******************************************************************\n"""
+    msg += '\n'.join(repo for repo in repos)
+    prompt = "\nWould you like to permanently enable these repositories? [y/N]: "
```
Comment (awood) [inline src/plugins/search-disabled-repos.py]: How would you feel about putting the enabled repositories in the block marked by asterisks?  If I'm reading correctly, right now the text would be

```
*******************************************************************
Dependency resolving was successful thanks to enabling these repositories:
*******************************************************************
foobar
foobar-optional
Would you like to permanently enable these repositories? [y/N]:
```

With something like

```
msg = """*******************************************************************
Dependency resolving was successful thanks to enabling these repositories:\n%s
*******************************************************************\n"""  %
"\n".join(repos)
```

We'd get

```
*******************************************************************
Dependency resolving was successful thanks to enabling these repositories:
foobar
foobar-optional
*******************************************************************
Would you like to permanently enable these repositories? [y/N]:
```

Which sets the prompt apart from the rest of the output.

**Comment by megaumi** [comment]
Thank you, Alex. Please see the new commits.

File: src/plugins/search-disabled-repos.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+# License and may only be used or replicated with the express permission of
+# Red Hat, Inc.
+
+import logging
+import os
+
+from yum.plugins import TYPE_CORE, TYPE_INTERACTIVE
+from yum.constants import TS_INSTALL_STATES
+
+requires_api_version = '2.7'
```
Comment (alikins) [inline src/plugins/search-disabled-repos.py]: This seems to be higher api version than what is available in rhel7?

File: src/plugins/search-disabled-repos.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+def is_repo_important(repo):
+    if repo.repofile != '/etc/yum.repos.d/redhat.repo':
+        return False
+    not_important_suffixes = ('debug-rpms', 'source-rpms', 'beta-rpms')
+    return not repo.id.endswith(not_important_suffixes)
+
+
+def suggest_enabling(conduit):
+    msg = conduit.pretty_output_restring() + """\n**********************************************************************
```
Comment (alikins) [inline src/plugins/search-disabled-repos.py]: How would be support translations for these strings?

File: src/plugins/search-disabled-repos.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+yum can be configured to try to resolve such errors by temporarily enabling
+disabled repos and searching for missing dependencies.
+To enable this functionality please set 'notify_only=0' in %s
+**********************************************************************\n""" % PLUGIN_CONF_PATH
+    conduit.info(1, msg) # yum's debuglevel 1
```
Comment (alikins) [inline src/plugins/search-disabled-repos.py]: With the default config, this goes to stdout whenever conduit.missing_requires is truthy? 

Just to verify, there will be no messages shown during normal use, but only with missing deps?

File: subscription-manager.spec [hunk 6]
```diff
@@ -296,6 +296,7 @@ rm -rf %{buildroot}
 # yum plugin config
 %config(noreplace) %attr(644,root,root) %{_sysconfdir}/yum/pluginconf.d/subscription-manager.conf
 %config(noreplace) %attr(644,root,root) %{_sysconfdir}/yum/pluginconf.d/product-id.conf
+%config(noreplace) %attr(644,root,root) %{_sysconfdir}/yum/pluginconf.d/search-disabled-repos.conf
```
Comment (alikins) [inline subscription-manager.spec]: Do we need Requires on newer versions of yum now? And conditional installation of the plugin and conf.

File: src/plugins/search-disabled-repos.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        for repo in helpful_new_repos:
+            # FIXME: replace with API calls
+            os.system('subscription-manager repos --enable=%s' % repo)
+            conduit.info(logging.DEBUG, 'Repo permanently enabled: %s' % repo)
+
+
+def is_repo_important(repo):
+    if repo.repofile != '/etc/yum.repos.d/redhat.repo':
+        return False
+    not_important_suffixes = ('debug-rpms', 'source-rpms', 'beta-rpms')
```
Comment (alikins) [inline src/plugins/search-disabled-repos.py]: candidate for config option?

Not sure if we will need config for '/etc/yum.repos.d/redhat.repo'. May be something to ask subman api for,

Comment (alikins) [inline src/plugins/search-disabled-repos.py]: (Note for my reference...). It would be useful if the Content info from rhsm/candlepin/subscription-manager knew which repos were "core" or "important". But that likely means new data from rcm, new schema for candlepin, new Certificate/Content classes for subman. Would be a lovely thing to be able to present to a user though.

File: src/plugins/search-disabled-repos.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    if not attempted:
+        return
+
+    used_repos = set(p.repoid for p in conduit.getTsInfo().getMembersWithState(output_states=TS_INSTALL_STATES))
+    helpful_new_repos = used_repos - old_enabled_repos
+
+    if prompt_permanently_enable_repos(conduit, helpful_new_repos):
+        for repo in helpful_new_repos:
+            # FIXME: replace with API calls
+            os.system('subscription-manager repos --enable=%s' % repo)
```
Comment (alikins) [inline src/plugins/search-disabled-repos.py]: Assuming a class/method like 
```
def default_enable_repo(repo):
     # do yum things to update the repo enabled flags 
     pass

def enable_repo(repo):
    repo_enabler = conduit.confString('main', 'repo_enable_tool', default = "builtin")
    if repo_enabler == "builtin":
         return default_enable_repo(repo)
    
    # popen/fork/whatever...
    run_repo_enabler(repo_enabler, repos)
```
If the important repo file list, the not_important_suffixes, and the repo_enable_tool were config
options, this plugin could be generic. And potentially apply to sat5 as well.

File: src/plugins/search-disabled-repos.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    if notify_only:
+        suggest_enabling(conduit)
+        return
+
+    if prompt_temporarily_enable_repos(conduit):
+        conduit.info(logging.DEBUG, 'Depsolving failed due to missing dependencies, temporarily enabling repos...')
+        global attempted
+        attempted = True
+        global old_enabled_repos
+        old_enabled_repos = set((repo.id for repo in repo_storage.listEnabled()))
```
Comment (alikins) [inline src/plugins/search-disabled-repos.py]: Does this handle temporarily disabled repos?  ie, 'yum install -y --disablerepo="rhel-7-server-rpms" something-with-deps'. Not sure what trumps what then.

File: src/plugins/search-disabled-repos.py (hunk 3 above)
Comment (megaumi) [inline src/plugins/search-disabled-repos.py]: That's correct, to make this functionality possible we are going to update the API in RHEL7.2. I posted the link to the scratch build of yum here https://bugzilla.redhat.com/show_bug.cgi?id=1232232#c17

File: src/plugins/search-disabled-repos.py (hunk 5 above)
Comment (megaumi) [inline src/plugins/search-disabled-repos.py]: That's right, unless there are missing deps and at least one disabled 'important' repo, there will be no messages.

File: src/plugins/search-disabled-repos.py (hunk 7 above)
Comment (megaumi) [inline src/plugins/search-disabled-repos.py]: I like the idea of introducing an option for not_important_suffixes.

Whether '/etc/yum.repos.d/redhat.repo' should be configurable is a question for subscription-manager team probably. I suspect there should be some API for that already?

File: src/plugins/search-disabled-repos.py (hunk 9 above)
Comment (megaumi) [inline src/plugins/search-disabled-repos.py]: In your example, rhel-7-server-rpms will get enabled by the plugin.

The idea behind this RFE is to make it easy for customers to resolve missing dependency problems. I believe the current implementation is consistent with this idea.

If we make not_important_suffixes configurable, it could be used to exclude 'rhel-7-server-rpms' from the list of repos to enable.

File: src/plugins/search-disabled-repos.py (hunk 4 above)
Comment (megaumi) [inline src/plugins/search-disabled-repos.py]: As I can see, plugins/subscription-manager.py isn't translated either?

File: subscription-manager.spec (hunk 6 above)
Comment (megaumi) [inline subscription-manager.spec]: With depending on yum API_VERSION Requires is probably not necessary. I'm not sure what you mean by the second question.

File: src/plugins/search-disabled-repos.py (hunk 3 above)
Comment (megaumi) [inline src/plugins/search-disabled-repos.py]: FYI I made a new yum scratch build supporting conduit.confList - https://bugzilla.redhat.com/show_bug.cgi?id=1232232#c19

File: src/plugins/search-disabled-repos.py (hunk 8 above)
Comment (megaumi) [inline src/plugins/search-disabled-repos.py]: This was discussed when the design of the RFE was discussed, and we do not want the plugin to be generic or work with sat5.

File: subscription-manager.spec (hunk 6 above)
Comment (alikins) [inline subscription-manager.spec]: Should be okay, assuming installing it on < RHEL 7.2 will just quietly skip the plugin if the requested API version isn't available.

If it shows warning or error on stdout/stderr, I'd prefer to make the plugin only installable with the correct yum version. (Or potentially, disabled by default in those cases).

File: src/plugins/search-disabled-repos.py (hunk 8 above)
Comment (alikins) [inline src/plugins/search-disabled-repos.py]: why?

File: src/plugins/search-disabled-repos.py (hunk 9 above)
Comment (alikins) [inline src/plugins/search-disabled-repos.py]: That may impact some product id install corner cases, but that's no big deal. 
(How? We try to track which rhsm/cdn repos are temporarily disabled to exclude them
from the list of inactive repos that are candidates for having product id certs deleted
post transaction. Very corner case...). Not a concern.

**Comment by alikins** [comment]
whats the best way to test this?

**Comment by alikins** [comment]
Ie, an example scenario where an install/update requires enabling extras/optional to resolve deps?

Are there particular packages/deps that often cause user confusion?

**Comment by alikins** [comment]
Related to previous question, this should only come into play when resolving deps right? I
shouldn't expect this to work:
```
    yum install --disablerepo="*" --enablerepo="rhel-7-server-rpms" install docker
``` 

But if a package installed from an enabled openshift channel that requires 'docker', I should expect extras to be enabled and used?

**Comment by alikins** [comment]
Attempting to test this with an example from the bug report.

```
# ghostscript was installed from rhel-7-server-rpms
# nodeps since this test was on my desktop and didn't want to remove all the deps
# (including redhat-lsb)
[subscription-manager (bz-1232232 u=)]$ sudo rpm -e --nodeps ghostscript

# Ie, only 'rhel-7-server-optional-rpms' enabled. No base rhel repo enabled. 
sudo yum --disablerepo="*" --enablerepo="rhel-7-server-optional-rpms" install ghostscript-devel
Loaded plugins: product-id, search-disabled-repos, subscription-manager
Resolving Dependencies
--> Running transaction check
---> Package ghostscript-devel.x86_64 0:9.07-18.el7 will be installed
--> Processing Dependency: ghostscript = 9.07-18.el7 for package: ghostscript-devel-9.07-18.el7.x86_64
--> Processing Dependency: libgs.so.9()(64bit) for package: ghostscript-devel-9.07-18.el7.x86_64
--> Finished Dependency Resolution
Error: Package: ghostscript-devel-9.07-18.el7.x86_64 (rhel-7-server-optional-rpms)
           Requires: libgs.so.9()(64bit)
Error: Package: ghostscript-devel-9.07-18.el7.x86_64 (rhel-7-server-optional-rpms)
           Requires: ghostscript = 9.07-18.el7
**********************************************************************
yum can be configured to try to resolve such errors by temporarily enabling
disabled repos and searching for missing dependencies.
To enable this functionality please set 'notify_only=0' in /etc/yum/pluginconf.d/search-disabled-repos.conf
**********************************************************************

Error: Package: ghostscript-devel-9.07-18.el7.x86_64 (rhel-7-server-optional-rpms)
           Requires: libgs.so.9()(64bit)
Error: Package: ghostscript-devel-9.07-18.el7.x86_64 (rhel-7-server-optional-rpms)
           Requires: ghostscript = 9.07-18.el7
 You could try using --skip-broken to work around the problem
** Found 8 pre-existing rpmdb problem(s), 'yum check' output follows:
ImageMagick-6.7.8.9-10.el7.x86_64 has missing requires of libgs.so.9()(64bit)
asciidoc-8.6.8-5.el7.noarch has missing requires of graphviz
cups-filters-1.0.35-16.0.0.1.el7.x86_64 has missing requires of libijs-0.35.so()(64bit)
ghostscript-cups-9.07-18.el7.x86_64 has missing requires of ghostscript = ('0', '9.07', '18.el7')
gutenprint-5.2.9-18.el7.x86_64 has missing requires of libijs-0.35.so()(64bit)
libspectre-0.2.7-4.el7.x86_64 has missing requires of libgs.so.9()(64bit)
netpbm-progs-10.61.02-9.el7.x86_64 has missing requires of ghostscript
redhat-lsb-printing-4.1-27.el7.x86_64 has missing requires of /usr/bin/gs

# edit search-disabled-repos.conf and flip notify_only=0
[subscription-manager (bz-1232232 u=)]$ sudo yum --disablerepo="*" --enablerepo="rhel-7-server-optional-rpms" install ghostscript-devel
Loaded plugins: product-id, search-disabled-repos, subscription-manager
Resolving Dependencies
--> Running transaction check
---> Package ghostscript-devel.x86_64 0:9.07-18.el7 will be installed
--> Processing Dependency: ghostscript = 9.07-18.el7 for package: ghostscript-devel-9.07-18.el7.x86_64
--> Processing Dependency: libgs.so.9()(64bit) for package: ghostscript-devel-9.07-18.el7.x86_64
--> Finished Dependency Resolution
Error: Package: ghostscript-devel-9.07-18.el7.x86_64 (rhel-7-server-optional-rpms)
           Requires: libgs.so.9()(64bit)
Error: Package: ghostscript-devel-9.07-18.el7.x86_64 (rhel-7-server-optional-rpms)
           Requires: ghostscript = 9.07-18.el7
**********************************************************************
Dependency resolving failed due to missing dependencies.
Some repositories on your system are disabled, but yum can enable them
and search for missing dependencies. This will require downloading
metadata for disabled repositories and may take some time and traffic.
**********************************************************************

Enable all repositories and try again? [y/N]: y
rhel-7-server-extras-rpms/x86_64                                                               | 2.9 kB  00:00:00     
rhel-7-server-extras-rpms/x86_64/updateinfo                                                    |  19 kB  00:00:00     
rhel-7-server-extras-rpms/x86_64/primary_db                                                    |  39 kB  00:00:00     
rhel-7-server-fastrack-rpms/x86_64                                                             | 2.9 kB  00:00:00     
rhel-7-server-fastrack-rpms/x86_64/updateinfo                                                  | 4.6 kB  00:00:00     
rhel-7-server-fastrack-rpms/x86_64/primary_db                                                  | 110 kB  00:00:00     
https://cdn.redhat.com/content/dist/rhel/server/7/7Server/x86_64/openstack-tools/7.0/os/repodata/repomd.xml: [Errno 14] HTTPS Error 404 - Not Found
Trying other mirror.
To address this issue please refer to the below knowledge base article 

https://access.redhat.com/articles/1320623

If above article doesn't help to resolve this issue please open a ticket with Red Hat Support.



 One of the configured repositories failed (Red Hat OpenStack Tools 7.0 for Red Hat Enterprise Linux 7 Server (RPMs)),
 and yum doesn't have enough cached data to continue. At this point the only
 safe thing yum can do is fail. There are a few ways to work "fix" this:

     1. Contact the upstream for the repository and get them to fix the problem.

     2. Reconfigure the baseurl/etc. for the repository, to point to a working
        upstream. This is most often useful if you are using a newer
        distribution release than is supported by the repository (and the
        packages for the previous distribution release still work).

     3. Disable the repository, so yum won't use it by default. Yum will then
        just ignore the repository until you permanently enable it again or use
        --enablerepo for temporary usage:

            yum-config-manager --disable rhel-7-server-openstack-7.0-tools-rpms

     4. Configure the failing repository to be skipped, if it is unavailable.
        Note that yum will try to contact the repo. when it runs most commands,
        so will have to try and fail each time (and thus. yum will be be much
        slower). If it is a very temporary problem though, this is often a nice
        compromise:

            yum-config-manager --save --setopt=rhel-7-server-openstack-7.0-tools-rpms.skip_if_unavailable=true

failure: repodata/repomd.xml from rhel-7-server-openstack-7.0-tools-rpms: [Errno 256] No more mirrors to try.
https://cdn.redhat.com/content/dist/rhel/server/7/7Server/x86_64/openstack-tools/7.0/os/repodata/repomd.xml: [Errno 14] HTTPS Error 404 - Not Found
```

In this case, enabling all the repos doesn't make sense. 

But the plugin did do what it's meant to do.

**Comment by alikins** [comment]
Let's merge it and sort it out later.

File: subscription-manager.spec (hunk 6 above)
Comment (megaumi) [inline subscription-manager.spec]: It does show a warning: "Plugin "search-disabled-repos" requires API 2.7. Supported API is 2.6.".

If you really don't want that warning you can require a specific yum version, but yum versions providing this API version will be different on RHEL7 and RHEL6.

If you disable the plugin by default, how would you handle enabling it when the correct yum version gets installed?

**Comment by megaumi** [comment]
> Related to previous question, this should only come into play when resolving deps right? I
> shouldn't expect this to work:

>   yum install --disablerepo="*" --enablerepo="rhel-7-server-rpms" install docker
> But if a package installed from an enabled openshift channel that requires 'docker', I should expect 
> extras to be enabled and used?

Yes, that's correct.
